src/backend/cdb/cdbquerycontextdispatching.c

```diff
@@ -102,7 +102,7 @@ int QueryContextDispatchingSizeMemoryLimit = 100 * 1024; /* KB */
 
 enum QueryContextDispatchingItemType
 {
-    MasterXid, TablespaceLocation, TupleType, EmptyTable, FileSystemCredential
+    MasterXid, TablespaceLocation, TupleType, EmptyTable, FileSystemCredential, Namespace
```

Comment:
So these enums identify portions of the `buffer` in `QueryContextInfo`. When the segment receives the data it checks for the presence of the enums and then rebuilds the data accordingly. 

The issue with dispatching the namespace from master to segment to pxf was two fold:
1. Master needs to serialize the data, segment needs to deserialize it
2. Segement needs to hold the `dfs_address` in a variable until the moment it actually invokes the pxf udf

The reason I added the namespace to the URL was so that I wouldn't need to add a new field to the `PG_FUNCTION_ARGS`.

I agree that adding namespace to URL is a bit of a hack, and this whole implementation is really the shortest path to get the namespace from master to segment so it definitely needs improvement.